### ISSUE TITLE
Feature Request: Allow Setting encoder to None When Using Pre-Computed Embeddings

### ISSUE THREAD

**Issue by mjaniec2013** [x-tabdeveloping/turftopic#45]
Currently, the KeyNMF constructor requires the encoder argument to be specified as a string:

```
keynmf_model = KeyNMF(
    n_components=num_topics,
    top_n=top_n,
    encoder='sentence-transformers/all-mpnet-base-v2',
    vectorizer=CountVectorizer(min_df=1)
)
```

This triggers the encoder model download via SentenceTransformers, even when embeddings are pre-computed and used later during model fitting:

```
sent_trans_model = SentenceTransformer("sentence-transformers/all-mpnet-base-v2")
embeddings = sent_trans_model.encode(texts)

keynmf_model.fit(texts, embeddings)
```

Loading the encoding model each time the constructor is called, especially within a loop (e.g., for topic number optimization), becomes inefficient and time-consuming.

Suggestion: Allow setting the encoder parameter to None in the constructor when using pre-computed embeddings. This would prevent unnecessary downloads and streamline the workflow for users who prefer to compute embeddings once and reuse them.

**Comment by x-tabdeveloping** [comment]
Hey @mjaniec2013 . Thanks for taking your time to write this. The issue is that KeyNMF also needs to learn and embed the vocabulary of the corpus, and that's why it requires the embedding model. There is an update on the way (very possibly this weekend or next week), where you will be able to extract the keywords with KeyNMF once, and then the vocab doesn't need to be encoded on the spot, I will make sure to comment on this issue when that happens.

A sensible way to deal with your problem for now would be to preload the encoder model in your code, and then use the `SentenceTransformer` object itself instead of the string you're now passing to KeyNMF.
Here's an example:
```python
sent_trans_model = SentenceTransformer("sentence-transformers/all-mpnet-base-v2")
embeddings = sent_trans_model.encode(texts)

for n in range(10, 50, 10):
    keynmf_model = KeyNMF(n_components=n, encoder=sent_trans_model)
    keynmf_model.fit(texts, embeddings=embeddings)
    ...
```
This way the sentence encoder does not need to be loaded on each run. Like I said, this process will be made much more streamlined in the next release. Tell me if I can be of help in the meanwhile.

**Comment by mjaniec2013** [comment]
Hello,

Thank you very much for the answer. As you suggested, passing a model instead of the string solves the issue of repeating model loading at initialization, resulting in 20% time savings for the optimization loop.

**Comment by mjaniec2013** [comment]
While the issue is solved, I've reopened the issue while waiting for the info on the upcoming update mentioned in the response.

**Comment by x-tabdeveloping** [comment]
The update came out yesterday! In Turftopic 0.4.0 you can precompute keyword matrices with a KeyNMF model and then use them later in training.

```python
from turftopic import KeyNMF
from sentence_transformers import SentenceTransformer

encoder = SentenceTransformer("sentence-transformers/all-mpnet-base-v2")

extractor_model = KeyNMF(
    # It doesn't matter what you choose, only comes into effect when calling fit()
    n_components=10,
    # You should probably choose a larger N than needed.
    # If the models you train afterwards has lower top_n, it will prune the keyword matrix anyway
    top_n=30,
    encoder=encoder,
)
# Here we extract the keywords
keyword_matrix: dict[str, float] = extractor_model.extract_keywords(texts)

for n in range(10, 50, 10):
    keynmf_model = KeyNMF(
          n_components=n,
          # This is no longer strictly necessary, the keyword matrix already contains all of the needed information
          encoder=encoder
    )
    keynmf_model.fit(keyword_matrix)
```

This is even faster than precomputing the embeddings, because the vocabulary and its embeddings also get learned and the only thing that has to happen inside the loop is the matrix decomposition step.

**Comment by mjaniec2013** [comment]
I've updated turftopic module to 0.4.3.
```

pip show turftopic

Name: turftopic
Version: 0.4.3
Summary: Topic modeling with contextual representations from sentence transformers.
Home-page:
Author: Márton Kardos
```

Unfortunately, the fit method does not have the keywords argument, allowing using a pre-computed documents-keywords list. I'm able to derive it from the input texts, but cannot use it as you suggested:

```
tokens_joined = [" ".join(token) for token in tokens]

keynmf_model = (KeyNMF(n_components=num_topics,
top_n=top_n,
encoder=sent_trans_model, # 'sentence-transformers/all-mpnet-base-v2',
vectorizer=CountVectorizer(min_df=1)))
# fit(tokens_joined, embeddings=embeddings))

if keywords_matrix is None:

keywords_matrix = keynmf_model.extract_keywords(tokens_joined)

keynmf_model = keynmf_model.fit(keywords=keywords_matrix)

keynmf_topics = keynmf_model.get_topics()

topics = reformat_topics(keynmf_topics)
```

Error:

keynmf_model = keynmf_model.fit(keywords=keywords_matrix)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: ContextualModel.fit() got an unexpected keyword argument 'keywords'

**Comment by x-tabdeveloping** [comment]
Sorry about that. I fixed this yesterday in 0.4.4.
Beware that you also have to pass something for the corpus when fitting the model, as scikit-learn requires it.

```python
model.fit(None, keywords=keywords)
```

**Comment by mjaniec2013** [comment]
It's me who should be sorry for taking your time :)
Nevertheless, the optimization function looks to work as intended now - thank you!

```
    def _get_coherence_keynmf(self, num_topics : int) -> float:

        # global tokens, corpus, dictionary, embeddings, top_n, sent_trans_model, keywords_matrix

        tokens_joined = [" ".join(token) for token in self.tokens]

        keynmf_model = KeyNMF(n_components=num_topics,
                              top_n=self.top_topic_keywords,
                              encoder=self.sent_trans_model,  # 'sentence-transformers/all-mpnet-base-v2',
                              vectorizer=CountVectorizer(min_df=1))

        if self.keywords_matrix is None:
            self.keywords_matrix = keynmf_model.extract_keywords(tokens_joined)

        keynmf_model = keynmf_model.fit(keywords=self.keywords_matrix)

        keynmf_topics = keynmf_model.get_topics()

        topics = self._reformat_topics(keynmf_topics)

        coherence_model = CoherenceModel(topics=topics,
                            texts=self.tokens,
                            dictionary=self.dictionary,
                            topn=self.top_topic_keywords,
                            coherence='c_v',
                            processes=-1)

        return coherence_model.get_coherence()
```

The function / method is additionally cached to eliminate redundant calculations in the optimization lopp.
